src/manifest.rs

```diff
@@ -33,15 +33,212 @@ use crate::{
+            Self::Unknown { .. } => Edition::Unknown,
+        }
+    }
+}
+
+mod serializer {
+    use super::*;
+    use serde::{ser::SerializeStruct, Serializer};
+
+    impl Serialize for Edition {
+        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+        where
+            S: Serializer,
+        {
+            match self {
+                Self::Canary => serializer.serialize_str(CANARY_EDITION),
+                Self::Unknown => serializer.serialize_str("unknown"),
+            }
+        }
+    }
+
+    impl Serialize for RawManifest {
+        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+        where
+            S: Serializer,
+        {
+            match *self {
+                RawManifest::Canary {
+                    ref package,
+                    ref dependencies,
+                } => {
+                    let mut s = serializer.serialize_struct("Canary", 3)?;
+                    s.serialize_field("edition", CANARY_EDITION)?;
+                    s.serialize_field("package", package)?;
+                    s.serialize_field("dependencies", dependencies)?;
+                    s.end()
+                }
+                RawManifest::Unknown {
+                    ref package,
+                    ref dependencies,
+                } => {
+                    let mut s = serializer.serialize_struct("Unknown", 2)?;
+                    s.serialize_field("package", package)?;
+                    s.serialize_field("dependencies", dependencies)?;
+                    s.end()
+                }
+            }
+        }
+    }
+}
+
+mod deserializer {
+    use serde::{
+        de::{self, MapAccess, Visitor},
+        Deserializer,
+    };
+
+    use super::*;
+
+    impl<'de> Deserialize<'de> for Edition {
+        fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+        where
+            D: Deserializer<'de>,
+        {
+            struct EditionVisitor;
+
+            impl<'de> serde::de::Visitor<'de> for EditionVisitor {
+                type Value = Edition;
+
+                fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                    formatter.write_str("a valid edition string")
+                }
+
+                fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+                where
+                    E: serde::de::Error,
+                {
+                    match value {
+                        c if c == CANARY_EDITION => Ok(Edition::Canary),
+                        _ => Ok(Edition::Unknown),
+                    }
+                }
+            }
+
+            deserializer.deserialize_str(EditionVisitor)
+        }
+    }
+
+    impl<'de> Deserialize<'de> for RawManifest {
+        fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+        where
+            D: Deserializer<'de>,
+        {
+            static FIELDS: &[&str] = &["package", "dependencies"];
```

Comment:
Design choice: Should this contain `edition`?

Reply:
I believe it should.